lib/ukdebug: Crash screen improvements

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This is a rework of #1114 with re-arch conventions in mind. It effectively decouples the crash screen logic from the rest of the kernel by implementing it as an exception handler provided by `libukdebug`.

Depends-on: #1527
Deprecates: #1114

michpappas committed:
uk/arch: Add support for nested exception handling

uknofault uses an exception event handler to handle inaccessible memory
areas. Using uknofault within exceptions would thus result into a stack
corruption for the outer exception. Introduce a set of functions that
allow disabling / re-enabling exception stack switch.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>
Signed-off-by: Marco Schlumpp <marco@unikraft.io>
Co-authored-by: Marco Schlumpp <marco@unikraft.io>

michpappas committed:
plat/common/x86: Adapt exception handling for crashdump

Introduce UK_EVENT_UNHANDLED_EXCEPTION, invoked by the fallback
trap handler, do_unhandled_trap(). This enables registering a
a custom crashdump handler. The default implementation is on
libukdebug.

Remove explicit page fault handler, as with the addition of the
unhandled exception event it no longer provides additional value
compared to the handler generated by DECLARE_TRAP_EC(). Instead,
pass the value of CR with the potentially faulting address to the
trap context.

Amend the trap context with a human-readable description of the
exception and a system error. The system error should be populated
by handlers.

Remove old register dump.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

jschlumpp committed:
plat/*: Remove old trace helpers

With the new crash screen they are now unused.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

michpappas committed:
lib/ukdebug: Add crash handling facilities

Introduce UK_CRASH_EX() as a macro that allows triggering a crash
from a location different from where the error originated, while
preserving the original execution context.

Implement default crash screen. The crash screen is registered as
a handler of the latest priority for the UNHANDLED_EXCEPTION event,
to provide a crashdump before executing a predefined action. The
default action is shutdown, and reboot is supported conditionally
to CONFIG_HAVE_REBOOT.

Provide UK_CRASH_EVENT, raised after a crashdump to give libraries
a chance to act after a crash. This separate event is required due
to UK_CRASH_EX() that bypasses the normal exception route.

Checkpatch-Ignore: LONG_LINE
Checkpatch-Ignore: QUOTED_WHITESPACE_BEFORE_NEWLINE
Signed-off-by: Michalis Pappas <michalis@unikraft.io>
Co-authored-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>
Co-authored-by: Marco Schlumpp <marco@unikraft.io>

michpappas committed:
plat/kvm: Select CONFIG_HAVE_REBOOT on EFI

The EFI implementation provides reboot support, so select the
HAVE_REBOOT option.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/common/arm64: Adapt exception handling to crashdump

Decouple the concept of exceptions from traps and move exception
definitions into common headers so that they can be referenced by
handling code.

Amend the trap context with a human-readable description of the
exception and a system error number populated by handlers.

Add generic unhandled exception event, invoked if nothing handles
the exception successfully. This allows last resort handling such
as a crashdump handler with debug info.

Remove old register dump code from the exception handler. Debug
info must now be provided by the crashdump handler.

Do not issue an event in the invalid exception handler. These
are exceptions taken frotm a different EL than the one Unikraft
is running on, so they should not be expected to happen under
normal circumstances.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukvmem: Populate trap system error

Populate the trap context's system error when unable
to handle a page fault.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat: Add HAVE_REBOOT to Config.uk

Invisible option to be selected by platforms or boot protocols
that provide reboot support.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on lib/ukdebug/Config.uk:
```diff
@@ -12,6 +12,61 @@ config LIBUKDEBUG_ENABLE_ASSERT
+	default y
+	depends on LIBUKPRINT
+	help
+		Output information on the kernel console if a crash occurred.
```
~~I know this commit is old and ported. I think the other configs here use correct indentation for the help message. We should update these as well.~~ (The new empty line added at the end might also be unnecessary) 

EDIT: Actually there may have been a confusion. We might actually indeed use tabs in this case, unlike LK KConfig styling. Resolving this and let's leave as is for now.

mogasergiu commented on lib/ukdebug/crash.c:
```diff
@@ -0,0 +1,138 @@
+{
+	struct uk_event_crash_parameter param;
+	struct ukarch_trap_ctx *ctx;
+#ifdef UKPLAT_LCPU_MULTICORE
```
I don't think this macro exists (or ever existed before). Maybe we should just delete related code for now?

michpappas replied:
Wow I'm impressed how this survived that long. That directive was introduced speculatively very early along with `plat/`  (828efb76ebb235966a52c9c7f7b76ef3bec36ec6) and parasitically lived in parallel to `HAVE_SMP` when multicore was introduced, before anyone noticed, at which point it started being removed. I'll replace with `HAVE_SMP`.

mogasergiu commented on lib/ukdebug/crash.c:
```diff
@@ -0,0 +1,138 @@
+	}
+#endif /* UKPLAT_LCPU_MULTICORE */
+
+#ifdef CONFIG_LIBUKDEBUG_CRASH_SCREEN
```
`#if`?

mogasergiu commented on lib/ukdebug/crashdump.c:
```diff
@@ -0,0 +1,72 @@
+void cdmp_gen_print_stack(unsigned long sp)
+{
+	char buf[64];
+	size_t ul_len = sizeof(unsigned long);
```
`__sz`

mogasergiu commented on lib/ukdebug/exportsyms.uk:
```diff
@@ -1,2 +1,4 @@
+_uk_crashdumpk
```
Does this symbol exist anymore ?

mogasergiu commented on lib/ukdebug/include/uk/crash.h:
```diff
@@ -0,0 +1,90 @@
+#endif
+
+enum uk_crash_reason {
+	UK_CRASH_REASON_INVALID,
```
Since this is in `uk/debug/`, wondering if all of these should have been namespaced as `UK_DEBUG_CRASH` or maybe the header should have instead been `uk/crash.h` like is the case for `uk/trace.h` and `uk/assert.h`

michpappas replied:
So this generally against our conventions, but I agree it makes sense to make an exception here give the ubiquitous nature of `libukdebug`. Will do 👍🏼

mogasergiu commented on lib/ukdebug/include/uk/debug/crash.h:
```diff
@@ -0,0 +1,90 @@
+	/**
+	 * Contains the errno code the systems exits with.
+	 */
+	int errno;
```
Let's not forget that dumb musl name clash :sneezing_face:

michpappas replied:
Adding some context here as that caught me by surprise. Apparently musl defines `errno` as a macro: https://git.musl-libc.org/cgit/musl/tree/include/errno.h#n16. Will switch to a different name.

mogasergiu commented on lib/ukdebug/include/uk/debug/crash.h:
```diff
@@ -0,0 +1,90 @@
+	__u64 arg3;
+};
+
+struct uk_event_crash_parameter {
```
I would name this as `uk_crash_...`

mogasergiu commented on lib/ukdebug/Makefile.uk:
```diff
@@ -4,15 +4,26 @@ $(eval $(call addlib_paramprefix,libukdebug,debug))
+LIBUKDEBUG_SRCS-$(CONFIG_LIBUKDEBUG_CRASH_SCREEN) += $(LIBUKDEBUG_BASE)/arch/$(CONFIG_UK_ARCH)/dumpsup.c
+LIBUKDEBUG_SRCS-$(CONFIG_LIBUKDEBUG_CRASH_SCREEN) += $(LIBUKDEBUG_BASE)/crashdump.c
+
+LIBUKDEBUG_SRCS-y += $(LIBUKDEBUG_BASE)/crash.c # FIXME conditional?
```
I believe we could revise this FIXME? Also wondering if we should do `|isr` for everything crash related... or maybe it doesn't matter as we are supposedly crashing anyway...

michpappas replied:
Actually that file must be unconditionally compiled as it manages `UK_EVENT_UNHANDLED_EXCEPTION`, which involves actions taken regardless `CONFIG_LIBUKDEBUG_CRASH_SCREEN`, as well as `UK_CRASH()` / `UK_CRASH_EX()`. I'll just remove the comment.

mogasergiu commented on arch/x86/x86_64/include/uk/asm/traps.h:
```diff
@@ -1,33 +1,8 @@
- * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
- * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
- * POSSIBILITY OF SUCH DAMAGE.
+/* Copyright (c) 2021, Karlsruhe Institute of Technology. All rights reserved.
```
Should this commit and its ARM equivalent actually come before the crash dump commit in the history? Saying this as the latter already uses `uk_errno` but it's only introduced afterwards.

michpappas replied:
Sure I can reorder the commits

mogasergiu commented on lib/ukdebug/include/uk/debug/crash.h:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Event for crashes. Will receive a uk_crash_event_param as data argument.
+ */
+#define UK_CRASH_EVENT uk_crash_event
```
I am not sure I understand the need for such a macro (I know it's from an old version of this work).

michpappas replied:
I think it's there to document the event in the public API, as you can't use `UK_EVENT()` in a header you include in more than one files.

mogasergiu commented on lib/ukdebug/arch/arm64/include/uk/arch/crash.h:
```diff
@@ -0,0 +1,32 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```
I know this patch set is old. But IIRC we are trying to get rid of asm directories for header inclusion paths

michpappas replied:
Good catch, will replace with `arch/`

mogasergiu commented on plat/kvm/arm/exceptions.S:
```diff
@@ -81,6 +83,12 @@
 	 */
 	msr     tpidrro_el0, x0
 
+	ldr	x0, =ukarch_except_switch_stack
```
Should've this been `ur_ldr` instead?
Also nit: I believe you could replace `cmp` + `beq` with a `cbz`:
```
ldr x0, =ukarch_except_switch_stack
cbz x0, 0f                    // Compare and Branch if Zero
```
It could help with branch prediction as this check is, although important, useless unless we are crashing.

michpappas replied:
~~This file contains several `ldr`, so I would rather not change one of them arbitrarily. If it needs to be converted to use `ldr_ur` that is a separate task.~~ **Edit:** Sorry, I didn't understand that the suggestion on `ur_ldr` was about literal pools, this is now fixed.

mogasergiu commented on plat/kvm/x86/traps.c:
```diff
@@ -151,19 +151,77 @@ static void tss_init(__lcpuidx idx)
+
+#define IDT_IST_SAVE_LEN 32
+/* Space for the IST values of all exception vectors */
+static __u8 idt_ist_saved[IDT_IST_SAVE_LEN];
```
Should this also be per CPU?

mogasergiu commented on plat/kvm/x86/traps.c:
```diff
@@ -151,19 +151,77 @@ static void tss_init(__lcpuidx idx)
+	if (*disable_nesting++)
+		return;
+
+	idt = ukplat_per_lcpu_current(cpu_idt);
```
This might prove to be unsafe as you cannot know for sure if you can safely access the lcpu idx field, given that it is an arbitrary section that does not guarantee the state of gs_base. Perhaps we could use `lcpu_get_current_in_except` to get the index and then reference the variable through `ukplat_per_lcpu`.

mogasergiu commented on plat/common/include/x86/traps.h:
```diff
@@ -108,23 +86,23 @@ static inline int _raise_event_##event(int trapnr, struct __regs *regs,	\
 		uk_pr_crit("trap handler returned error: %d\n", rc);	\
 									\
-	if (!rc)							\
+	if (rc == UK_EVENT_NOT_HANDLED)					\
```
Should this be `unlikely(rc == UK_EVENT_NOT_HANDLED || rc < 0)` (ditto below)?

michpappas replied:
There is a separate check for `rc` right above

mogasergiu commented on plat/common/arm/traps_arm64.c:
```diff
@@ -164,6 +165,21 @@ static enum aarch64_trap esr_to_trap(__u64 esr)
+
+void ukarch_pop_nested_exceptions(void)
+{
+	--ukplat_per_lcpu_current;
```
uhh, I think you are fixing this by mistake in the next commit :sweat_smile:

mogasergiu commented on lib/ukdebug/crash.c:
```diff
@@ -0,0 +1,153 @@
+	crash_shutdown();
+}
+
+int uk_debug_crash_handler(void *data)
```
Should this be `uk_crash_handler` instead and `static`?

mogasergiu commented on lib/ukdebug/crashdump.h:
```diff
@@ -0,0 +1,37 @@
+/* Must be implemented in the architecture-specific ukdebug support */
+void cdmp_arch_print_registers(struct __regs *regs);
+void cdmp_arch_print_stack(struct __regs *regs);
+#if !__OMIT_FRAMEPOINTER__
```
Should this also be guarded by `CONFIG_LIBUKDEBUG_CRASH_PRINT_CALL_TRACE`?

mogasergiu commented on lib/ukdebug/include/uk/crash.h:
```diff
@@ -0,0 +1,115 @@
+};
+
+/**
+ * Event raised after a crashdump to give libraries a chance
```
Ok so I guess we will keep this.
This might mean that when we touch already existing (or add new) events, we should also add a similar macro with docs for them, as there are several that don't do this.

mogasergiu commented on lib/ukdebug/Makefile.uk:
```diff
@@ -4,15 +4,26 @@ $(eval $(call addlib_paramprefix,libukdebug,debug))
+LIBUKDEBUG_SRCS-$(CONFIG_LIBUKDEBUG_CRASH_SCREEN) += $(LIBUKDEBUG_BASE)/arch/$(CONFIG_UK_ARCH)/dumpsup.c
+LIBUKDEBUG_SRCS-$(CONFIG_LIBUKDEBUG_CRASH_SCREEN) += $(LIBUKDEBUG_BASE)/crashdump.c
+
+LIBUKDEBUG_SRCS-y += $(LIBUKDEBUG_BASE)/crash.c
```
since this uses inclusions that are guarded by `LIBUKDEBUG_CRASH_SCREEN`, maybe it should also be guarded by it? In which case, its `#if CONFIG_LIBUKDEBUG_CRASH_SCREEN` could also be removed.
(Btw I just noticed but what is the `LIBUKDEBUG_CINCLUDES-y  += -I$(LIBUKDEBUG_BASE)` above for?)

michpappas replied:
`crash.c` contains behavior that is not limited to the crash screen and most importantly the implementation of `UK_CRASH()` so it should not be compiled conditionally

michpappas replied:
>  (Btw I just noticed but what is the LIBUKDEBUG_CINCLUDES-y  += -I$(LIBUKDEBUG_BASE) above for?)

Yes, that is a leftover from the original patches, thanks for spotting.

mogasergiu commented on lib/ukdebug/arch/arm64/dumpsup.c:
```diff
@@ -0,0 +1,67 @@
+		fp = frame[0];
+	}
+}
+#endif /* !__OMIT_FRAMEPOINTER__ && CONFIG_LIBUKNOFAULT */
```
I think you meant `CONFIG_LIBUKDEBUG_CRASH_PRINT_CALL_TRACE && !__OMIT_FRAMEPOINTER__`
(Ditto x86)

mogasergiu commented on lib/ukdebug/arch/x86_64/include/uk/arch/crash.h:
```diff
@@ -0,0 +1,32 @@
+
+extern __bool _uk_crash_explicit;
+
+#define uk_crash_trigger()						\
```
Shouldn't this be `ukarch_crash_trigger`? Same for the header guards, `__UK_ARCH_CRASH_H__`?
(Ditto x86)

michpappas replied:
No, we don't have such a convention

michpappas committed:
plat/xen: Stubs nested exception management func

Functions for nested exception management are now
expected to be implemented by all platforms. Xen
operates on a single stack, so implement these
operations as stubs.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>
Approved-by: Sergiu Moga <sergiu@unikraft.io>
